frontend/src/components/widgets/DataFrame/columns/utils.ts

```diff
@@ -0,0 +1,358 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  GridCell,
+  Theme as GlideTheme,
+  TextCell,
+  GridCellKind,
+  LoadingCell,
+  GridColumn,
+} from "@glideapps/glide-data-grid"
+import { toString, merge, isArray } from "lodash"
+import numbro from "numbro"
+
+import { Type as ArrowType } from "src/lib/Quiver"
+import { notNullOrUndefined, isNullOrUndefined } from "src/lib/utils"
+
+/**
+ * Interface used for defining the properties (configuration options) of a column.
+ * These options can also be used to overwrite from user-defined column config.
+ */
+export interface BaseColumnProps {
+  // The id of the column.
+  readonly id: string
+  // The title of the column.
+  readonly title: string
+  // The index number of the column.
+  readonly indexNumber: number
+  // The arrow data type of the column.
+  readonly arrowType: ArrowType
+  // If `True`, the column can be edited.
+  readonly isEditable: boolean
+  // If `True`, the column is hidden (will not be shown).
+  readonly isHidden: boolean
+  // If `True`, the column is a table index.
+  readonly isIndex: boolean
+  // If `True`, the column is a stretched.
+  readonly isStretched: boolean
+  // The initial width of the column
+  readonly width?: number
+  // Column time selected via column config
+  readonly customType?: string
+  // Additional metadata related to the column type.
+  readonly columnTypeMetadata?: Record<string, any>
+  // The content alignment of the column.
+  readonly contentAlignment?: "left" | "center" | "right"
+  // Theme overrides for this column.
+  readonly themeOverride?: Partial<GlideTheme>
+}
+
+/**
+ * The interface that is implemented by any column type.
+ */
+export interface BaseColumn extends BaseColumnProps {
+  readonly kind: string
+  // Defines the sort mode that should be used for this column type:
+  // default: Sorts by interpreting all values as strings.
+  // smart: Detects if value is a number or a string and sorts accordingly.
```

Comment:
Personally, when I was writing a column implementation, I think it's important to write down a note that the sorting is done through copyData.